src/server/pkg/work/work.go

```diff
@@ -117,106 +199,124 @@ type ProcessFunc func(context.Context, *Task, *Task) error
+		return taskQueue.runTask(ctx, taskID, func(taskEntry *taskEntry) {
+			if err := w.taskFunc(task, taskEntry, processFunc); err != nil && taskEntry.ctx.Err() != context.Canceled {
+				fmt.Printf("errored in task callback: %v\n", err)
+			}
+		})
 	})
 }
 
-func (w *Worker) processTask(ctx context.Context, task *Task) error {
-	var eg errgroup.Group
-	ctx, cancel := context.WithCancel(ctx)
-	// Watch for task termination.
-	// This will handle completion, failure, and cancellation.
-	eg.Go(func() error {
-		return w.taskCol.ReadOnly(ctx).WatchOneF(task.Id, func(e *watch.Event) error {
-			var id string
-			taskInfo := &TaskInfo{}
-			if err := e.Unmarshal(&id, taskInfo); err != nil {
+func (w *Worker) taskFunc(task *Task, taskEntry *taskEntry, processFunc ProcessFunc) error {
+	claimWatch, err := w.claimCol.ReadOnly(taskEntry.ctx).WatchOne(task.ID, watch.WithFilterPut())
+	if err != nil {
+		return err
+	}
+	subtaskWatch, err := w.subtaskCol.ReadOnly(taskEntry.ctx).WatchOne(task.ID, watch.WithFilterDelete())
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case e := <-claimWatch.Watch():
+			if e.Type == watch.EventError {
+				return e.Err
+			}
+			if e.Type != watch.EventDelete {
+				continue
+			}
+			var subtaskKey string
+			if err := e.Unmarshal(&subtaskKey, &Claim{}); err != nil {
 				return err
 			}
-			if taskInfo.State != State_RUNNING {
-				cancel()
-				return errutil.ErrBreak
+			if err := taskEntry.runSubtask(w.subtaskFunc(task, subtaskKey, processFunc)); err != nil {
+				return err
 			}
-			return nil
-		})
-	})
-	// Process the subtasks.
-	eg.Go(func() (retErr error) {
-		// (bryce) cleanup of subtasks needs to be re-thought.
-		// Cannot cleanup in this manner, because it messes up subtasks claims when a worker errors.
-		// Cleaning up in the master results with a race condition with workers
-		// waiting on work termination signal and subtask claims to expire.
-		//defer func() {
-		//	// Attempt to cleanup subtask entries.
-		//	if _, err := col.NewSTM(context.Background(), w.etcdClient, func(stm col.STM) error {
-		//		w.subtaskCol.ReadWrite(stm).DeleteAllPrefix(work.Id)
-		//		return nil
-		//	}); err != nil && retErr == nil {
-		//		retErr = err
-		//	}
-		//}()
-		if err := w.processSubtasks(ctx, task); err != nil {
-			return err
-		}
-		// Wait for a deletion event (ttl expired) before attempting to process the subtasks again.
-		return w.subtaskCol.ReadOnly(ctx).WatchOneF(task.Id, func(e *watch.Event) error {
-			return w.processSubtasks(ctx, task)
-		}, watch.WithFilterPut())
-	})
-	if err := eg.Wait(); err != nil {
-		// Work terminated.
-		if ctx.Err() == context.Canceled {
-			return nil
+		case e := <-subtaskWatch.Watch():
+			if e.Type == watch.EventError {
+				return e.Err
+			}
+			var subtaskKey string
+			if err := e.Unmarshal(&subtaskKey, &TaskInfo{}); err != nil {
+				return err
+			}
+			if err := taskEntry.runSubtask(w.subtaskFunc(task, subtaskKey, processFunc)); err != nil {
```

Comment:
So I think you should be able to skip this part if the subtask state isn't `RUNNING`.  I tried this with my benchmarks and it got slower, though, which is pretty weird.

Reply:
Yeah, that is true, although I don't think it really matters with the queuing model. You are basically avoiding two channel writes and an iteration through the tasks in the queue by doing this. I can still add this, but I question if it is worth the extra lines of code considering the task queue is not expected to have a very large number of tasks (so I would like to keep it as simple as possible). It does seem weird that this made your benchmark slower. I guess it could technically result with more etcd contention because you could have the subtask that is running and events from the claim watcher performing requests in parallel.